### PR TITLE
[QC][PPU] Add flash_attn_varlen_func_w8a8_int8

### PR DESCRIPTION
**Pull request by July-h5kf3**
### PR Category

Operator

### Type of Change

New Feature, Performance Optimization

### Description

Adds `flash_attn_varlen_func_w8a8_int8` for THead / PPU, dispatched by input dtype through the existing public `flaggems_vllm.flash_attn_varlen_func` entry. The implementation is grouped with the attention family in `src/flaggems_vllm/runtime/backend/_thead/fused/attention.py`, with dedicated test and benchmark files.

Q/K/V inputs are INT8. QK uses INT8 dot products with INT32 accumulation; softmax and online output accumulation use FP32. For paged KV with maximum query length >= 128, a Triton preprocessing kernel packs K/V contiguously by KV head and converts V to FP16. PV then uses FP16 operands with FP32 accumulation. Other paths use quantized INT8 probabilities and INT8 PV with explicit zero-point correction. Grouped query heads share K/V loads, and mixed batches use a compact query grid to reduce empty programs.

Supports packed and paged KV, MHA/GQA/MQA, head dimensions 64/128, FP16/BF16 output, LSE, causal/local masks, ALiBi, and softcap. Tests cover these paths, block-varying and broadcast descales, strided inputs, and empty KV. The benchmark reuses the existing FlagGems-vllm runner and all 18 workloads from vLLM's standard attention configuration. Operator metadata and CI test selection map the shared attention module to the dedicated tests and benchmark.

This is inference forward only. Dropout, attention-probability output, auxiliary queries, scheduler parameters, context parallelism, split-KV, and FA3 are unsupported. The long-query preprocessing workspace uses three bytes per physical K-cache element and packs all physical pages; its cost can increase when allocated cache capacity substantially exceeds the KV used by the call.

### Issue

Related NVIDIA implementation: https://github.com/flagos-ai/FlagGems-vllm/pull/749

### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [x] Change is fully covered by a UT.

### Performance

Measured on PPU-ZW810E with PyTorch 2.10.0, PPU vLLM 0.19.0+cu130, and FlagTree built from official source commit `4cd822da5e245d408536e378463dcde5939d6c7e`.

**Baseline: the existing FlagGems-vllm BF16 implementation of `flaggems_vllm.flash_attn_varlen_func`, running on the same device.** The benchmark's `_varlen_bf16_baseline` passes the original BF16 Q/K/V to the existing floating-point path in `ops/attention.py`. The INT8 implementation receives quantized versions of those inputs. Request lengths, page tables, head counts, head dimensions, and causal settings are identical. PPU vLLM's FA2/FA3 entry points do not provide a usable native INT8 baseline; the BF16 column is the FlagGems-vllm implementation, not native vLLM CUDA attention.

Timing uses the repository CUDA Graph benchmark with 5 warmup and 10 measurement iterations. All GPU work within the attention call is included, including K/V packing and V conversion. Input quantization is excluded. CUDA Graph timing excludes Python dispatch and host allocation overhead. Latencies below are arithmetic means of the recorded measurements; speedup is mean BF16 latency divided by mean INT8 latency.

Workloads follow [vLLM v0.19.0 standard_attention.yaml](https://github.com/vllm-project/vllm/blob/v0.19.0/benchmarks/attention_benchmarks/configs/standard_attention.yaml), with Hq/Hkv=32/8, D=128, page size=16, and causal=True. Each request group is `(request count, query length, total KV length)`; `+` denotes a mixed batch. Q shape is `[total_q, Hq, D]`; K/V shape is `[physical pages, page size, Hkv, D]`, including allocated unused pages.

|Request groups `(B, Sq, Sk)`|Q shape|K/V shape (each)|BF16 ms|INT8 ms|Speedup (BF16/INT8)|
|---|---|---|---:|---:|---:|
|`(1, 512, 512)`|`[512, 32, 128]`|`[32, 16, 8, 128]`|0.1094|0.0783|1.397x|
|`(1, 2048, 2048)`|`[2048, 32, 128]`|`[128, 16, 8, 128]`|0.9965|0.7456|1.337x|
|`(1, 4096, 4096)`|`[4096, 32, 128]`|`[256, 16, 8, 128]`|3.4513|2.6918|1.282x|
|`(1, 8192, 8192)`|`[8192, 32, 128]`|`[512, 16, 8, 128]`|12.5927|10.2357|1.230x|
|`(8, 1, 1024)`|`[8, 32, 128]`|`[512, 16, 8, 128]`|0.0853|0.0403|2.117x|
|`(16, 1, 2048)`|`[16, 32, 128]`|`[2048, 16, 8, 128]`|0.3349|0.1045|3.205x|
|`(32, 1, 1024)`|`[32, 32, 128]`|`[2048, 16, 8, 128]`|0.3452|0.1153|2.994x|
|`(64, 1, 4096)`|`[64, 32, 128]`|`[16384, 16, 8, 128]`|2.0346|0.9193|2.213x|
|`(2, 2048, 2048)` + `(8, 1, 1024)`|`[4104, 32, 128]`|`[1280, 16, 8, 128]`|3.4473|1.6458|2.095x|
|`(4, 1024, 1024)` + `(16, 1, 2048)`|`[4112, 32, 128]`|`[2560, 16, 8, 128]`|5.9015|1.1081|5.326x|
|`(2, 4096, 4096)` + `(32, 1, 1024)`|`[8224, 32, 128]`|`[8704, 16, 8, 128]`|12.5722|5.8606|2.145x|
|`(16, 2, 1024)`|`[32, 32, 128]`|`[1024, 16, 8, 128]`|0.5386|0.0538|10.020x|
|`(16, 4, 1024)`|`[64, 32, 128]`|`[1024, 16, 8, 128]`|0.5445|0.0540|10.092x|
|`(16, 8, 1024)`|`[128, 32, 128]`|`[1024, 16, 8, 128]`|0.5431|0.1139|4.768x|
|`(32, 4, 2048)`|`[128, 32, 128]`|`[4096, 16, 8, 128]`|2.0381|0.2495|8.170x|
|`(8, 8, 4096)`|`[64, 32, 128]`|`[2048, 16, 8, 128]`|1.1331|0.2022|5.604x|
|`(1, 1024, 2048)`|`[1024, 32, 128]`|`[128, 16, 8, 128]`|0.6693|0.5333|1.255x|
|`(2, 1024, 4096)`|`[2048, 32, 128]`|`[512, 16, 8, 128]`|2.6080|2.2552|1.156x|

**Overall: 3.689x arithmetic mean speedup over FlagGems-vllm BF16 across all 18 shapes.** Each shape has equal weight; the average is calculated from unrounded per-shape speedups. All **18/18** shapes are faster, with a minimum per-shape speedup of **1.156x**.

These results cover the listed single-attention workloads. They do not establish model-level accuracy, end-to-end serving throughput, or performance with substantially larger physical cache allocations.

### CODE REVIEW

**Comment by 0x45f** [comment]
please fix conflicts

**Comment by July-h5kf3** [comment]
> please fix conflicts

Done！

File: benchmark/test_flash_attn_varlen_func_w8a8_int8.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+                _varlen_int8(bf16_args, int8_args),
+                _varlen_bf16_baseline(reference_args, int8_args),
+                atol=0.03,
+                rtol=0.03,
+            )
+            yield bf16_args, int8_args
+
+
+def _varlen_bf16_baseline(bf16_args, int8_args):
+    return flaggems_vllm.flash_attn_varlen_func(*bf16_args[:-1], **bf16_args[-1])
```
Comment (0x45f) [inline benchmark/test_flash_attn_varlen_func_w8a8_int8.py]: please compare with vllm.flash_attn_varlen_func and show speedup

Comment (0x45f) [reply benchmark/test_flash_attn_varlen_func_w8a8_int8.py]: can we compare with `vllm.flash_attn_varlen_func(..., fa_version=3)` ?
